src/snowflake/snowpark/_internal/utils.py

```diff
@@ -669,3 +669,64 @@ def get_copy_into_table_options(
         elif k not in NON_FORMAT_TYPE_OPTIONS:
             file_format_type_options[k] = v
     return file_format_type_options, copy_options
+
+
+def convert_ids_to_upper_case(name: str) -> str:
+    """convert a fully qualfied name to upper case when needed. The double-quoted part won't be converted.
+    This function doesn't validate whether the name has the right pattern.
+    Instead, it assumes the name has the right pattern.
+    """
+    i = 0
+
+    def skip_space() -> None:
+        nonlocal i
+        while i < len(name) and name[i] == " ":
+            i += 1
+
+    def find_next_id() -> str:
```

Comment:
I think `find_next_id` and `i` should be wrapped using a [generator](https://wiki.python.org/moin/Generators).